File: Code/base/src/util/timer.rs
//! # Timer

use std::time::{Instant, Duration};

/// Timer struct that will keep track of the time spent between pings.
#[derive(Debug, Copy, Clone, PartialOrd, PartialEq)]
pub struct Timer {
    ctime: Instant,
    ptime: Instant,
    dt: Duration,
    target_dt: Option<Duration>
}

impl Timer {
    /// Creates a new `Timer` object with the given optional target delta time.
    pub fn new(target_dt: Option<Duration>) -> Self {
        Self {
            ctime: Instant::now(),
            ptime: Instant::now(),
            dt: Duration::new(0,0),
            target_dt,
        }
    }

    /// Pings the timer, returning the amount of time that has passed since the
    /// last ping, optionally waiting for the `target_dt` duration to pass.
    pub fn ping(&mut self) -> Duration {
        if let Some(target_dt) = self.target_dt {
            while Instant::now() - self.ptime < target_dt {}
        }

        self.ptime = self.ctime;
        self.ctime = Instant::now();

        self.dt = self.ctime - self.ptime;

        self.dt
    }
}

File: Code/base_cpp/src/util/Timer.cpp
#include "Timer.hpp"

namespace RanOS
{
    void Timer::Wait(Duration t) {
        auto ctime = TIME_NOW();
        while(TIME_NOW()-ctime < t) {}
    }

    Timer::Timer(Option<Duration> target):
        ctime(TIME_NOW()),
        ptime(TIME_NOW()),
        dt(),
        target_dt(target)
    {
    }

    Duration Timer::ping() {
        self.ptime = self.ctime;

        if(self.target_dt.is_some()) {
            auto t_dt = self.target_dt.unwrap();

            while((self.ctime - self.ptime) < t_dt) {
                self.ctime = TIME_NOW();
            }
        } else {
            self.ctime = TIME_NOW();
        }

        self.dt = self.ctime - self.ptime;
        return self.dt;
    }

    void Timer::reset() {
        self = Timer(self.target_dt);
    }
}
